### PR TITLE
Speed up sanitizer stack traces

### PR DESCRIPTION
**Pull request by quantum5**
### Rationale

Running the sqlite3 test with the experimental LSan code in #8711 results in around 12x slowdown compared to no LSan. We can do better. The obvious suspect is the stack tracing code, since that is called every `malloc`, and on native, the sanitizers take special effort in making stack unwinding fast, even at the cost of accuracy.

This PR contains most of the easy changes, and with it, the LSan code results in around 4x slowdown, a 3x improvement compared to before.

### Changes

* directly unwind into a C buffer in one call instead of calling a JS helper in a loop
* make `emscripten_stack_snapshot` return `__builtin_return_address(0)` since the value is readily available, saving the trouble of doing another stack trace
* use `new Error().stack` instead of `emscripten_get_callstack_js`, avoiding parsing and reformatting stack trace just to parse it again
* Look up the PC value in the source map directly (post-#8762, it's always the module offset) instead of parsing the stack frame again
* Cache the last result of `emscripten_pc_get_source_js` since `emscripten_pc_get_*` all call it, and typically with the same arguments

WASM library cache needs to be cleared after this change lands, since the old `libsanitizer_common_rt_wasm.a` depends on a function that is now removed.

### CODE REVIEW

File: src/library.js [hunk 1]
```diff
@@ -4734,30 +4738,55 @@ LibraryManager.library = {
   // JavaScript frames, so we have to rely on PC values. Therefore, we must be able to unwind from
   // a PC value that may no longer be on the execution stack, and so we are forced to cache the
   // entire call stack.
-  emscripten_stack_snapshot__deps: ['emscripten_get_callstack_js', 'emscripten_generate_pc', '$UNWIND_CACHE'],
+  emscripten_stack_snapshot__deps: ['emscripten_get_callstack_js', 'emscripten_generate_pc',
+                                    '$UNWIND_CACHE', '_emscripten_save_in_unwind_cache'],
   emscripten_stack_snapshot: function () {
-    var callstack = _emscripten_get_callstack_js(0).split('\n');
+    var callstack = new Error().stack.split('\n');
```
Comment (kripken) [inline src/library.js]: get_callstack_js does some fixing up of the stack trace. do we not need that to work around browser differences? or do we do the same workarounds here too (which is maybe not great if so)?

File: src/library.js [hunk 2]
```diff
@@ -4734,30 +4738,55 @@ LibraryManager.library = {
...
-      return 0;
+      stack = new Error().stack.split('\n');
+      if (stack[0] == 'Error') {
+        stack.shift();
+      }
+      __emscripten_save_in_unwind_cache(stack);
+    }
+
+    var offset = 2;
+    while (stack[offset] && _emscripten_generate_pc(stack[offset]) != addr)
```
Comment (kripken) [inline src/library.js]: `{ }` if a body is on another line

File: src/library.js (hunk 1 above)
Comment (quantum5) [reply src/library.js]: We are working around browser differences here by having separate regexes for v8 and spidermonkey. We used to have that anyways due to minor differences in how wasm stack frames are formatted in the two engines, despite `emscripten_get_callstack_js`'s attempt at unifying it. We might as well cut the middleman and not parse the results twice.

File: src/preamble.js [hunk 3]
```diff
@@ -1055,8 +1060,16 @@ function createWasm(env) {
 #endif
   }
 
+#if 'emscripten_generate_pc' in addedLibraryItems
+  addRunDependency('offset-converter');
```
Comment (kripken) [inline src/preamble.js]: Is this code in another PR too? I thought I already wrote a comment on this.

Comment (quantum5) [reply src/preamble.js]: Yes, this code came from #8762.

**Comment by quantum5** [comment]
The PR that this PR depended on, #8762, has been merged. The diff is now clean.

**Pull request merged**